### ISSUE TITLE
SQL logging in 1.7.244 seems entirely broken for me

### ISSUE THREAD

**Issue by ashdar** [PowershellFrameworkCollective/psframework#546]
I'm working on a system that will require logging hundreds (maybe thousands) of entries per day, for each execution of a PowerShell script. My goal is to use the SQL logging facility for this. Here is a bit of a backstory.

I've got dbatools 1.1.133 installed and I'm running PowerShell 7-x64.

I have code that works OK, using 1.6.214, until I try to log more than about 100 items in a single script. Then, I run into what seems to be a bug in 1.6.214 that causes connections to not be reused. If you look at line 35 in sql.provider.ps1, "if ($SqlInstance.ConnectionContext.IsOpen -ne 'True')" is testing the boolean return of a function IsOpen against a string 'true'). (I'm also a little concerned because that code seems to be using a server name/string and not a connection object of some sort, but one thing at a time.)

Now, I would like to put some more effort into this so that I could submit a patch, but that's an old version of PSFramework and the first thing I need to do get the latest version of PSFramework onto my system. BUT, if I upgrade to 1.7.244 and run any kind of test script, such as a script based on the [example here](https://psframework.org/documentation/documents/psframework/logging/loggingto/sql.html), absolutely nothing seems to happen. Nothing is logged, the logging tables are not created, I don't get login failure messages, there is no complaint about a missing database name or anything I would expect. There are only failures that can only be seen using "(Get-PSFLoggingProviderInstance).Errors", which I found in a different open issue. Here is what my repro script looks like:

```
$paramSetPSFLoggingProvider = @{
    Name         = 'sql'
    InstanceName = 'MyTask'
    SqlServer    = 'server.fqdn.com'
    Database     = 'LoggingDB'
    Table        = 'Logs'
    Enabled      = $true
}
Set-PSFLoggingProvider @paramSetPSFLoggingProvider

# Verbose
Write-PSFMessage -Message "Test Message"
```

Note that this example does not provide a correct/usable SqlServer, database or table name. It _should_ fail with some sort of connection or credential error. Instead, here is what the failure looks like:
![image](https://user-images.githubusercontent.com/11427665/193483061-8a4c6127-956e-44b1-bea0-f62e4ac35b8b.png)

That's a very odd error message because it seems like an object[] should be an object[]. It makes me think that something is calling Export-DataToSql in a very odd way. "Line 6" notwithstanding, I suspect that the offending code is really in the $message_event  script block down around line 229 of the sql.provider.ps1 file. 

I can't get VS Code to step into Write-PSFMessage. I have set a couple of dozen breakpoints inside of the PSFramework code, on any line that seems slightly related, but none of my breakpoints ever get hit. It _feels like_ the error happens before Write-PSFMessage even tries a database connection-no table was created to hold logged events and so forth. Since none of my breakpoints gets hit, I am guessing that either the function-pointer-esque code in the logging provider infrastructure code is preventing it somehow or it has something to do with the code in the cs file(s) (maybe my breakpoints won't get hit if the PS code is called from within the csharp code).

Can someone provide some advice? I can't debug this any further without some kind of revelation.

**Comment by FriedrichWeinmann** [comment]
Hi @ashdar,
thanks for finding and reporting this (and the excellent write-up!).

The error happens inside the logging provider code - debugging into Write-PSFMessage is futile, as it is not only written in C# ... it also just queues up the message into a queue, the logging provider picks up.
Which happens in a background runspace, so you get to see only little.
The provider source code is further used to dynamically generate modules, losing any relationship to the files the code came from.

Architecturally it was necessary, alas, so I had to provide a way to debug it anyway.
After configuring the logging provider, the first thing you need is the instance object:

```powershell
$instance = Get-PSFLoggingProviderInstance -ProviderName sql
```

We probably also need a message object to work with:

```powershell
Write-PSFMessage foo
$message = Get-PSFMessage | Select-Object -Last 1
```

You can now invoke the phases directly like this:

```powershell
# Initial Setup
& $instance.BeginCommand

# Start of a logging Cycle
& $instance.StartCommand

# Log a Message
& $instance.MessageCommand $message

# End of a logging Cycle
& $instance.EndCommand

# Shutdown sequence for the instance
& $instance.FinalCommand
```

You can get live error records of what is going wrong and should be able to utilize breakpoints using Set-PSBreakpoint.

**Comment by ashdar** [comment]
I just spent a while stepping through code. I'm confused.
In a plain old console window, away from Visual Studio Code, I can do this:
Set-PSBreakpoint -command "New-DefaultSqlDatabaseAndTable"

And then run my repro script-which does, in fact, stop. I can then examine the stack and start stepping forward, executing the begin{} of that function, line by line, all the way down and through the if ($Credential) check, then on the last } of the begin{}. The next attempt to step out and start stepping through the process{} loop goes wobbly. 

![image](https://user-images.githubusercontent.com/11427665/193628157-39282f53-d2b7-48b9-a5ec-b998a879f1fa.png)

The only code I see that looks like it would be working on Function: is down around PSFramework.psm1(15896), just below where the Get-ConfigValue function is defined.  I don't know how the program got there.  Worse, I get kicked out of the debugger, even though the [DBG] prompt is still there. 

Is there a relationship between the .Begin{} in New-DefaultSqlDatabaseAndTable and the .Begin, .Start, .Message and so forth-the different events for the Provider.  If I look at the definitions for the $begin_event for the other (non-SQL) providers, I don't see anything else that has Begin{}, Process{} and End{} script blocks like New-DefaultSqlDatabaseAndTable does. I am speculating that the code is accidentally using .Begin() to mean different things, somehow, and when my script tries to unwind the stack when returning from New-DefaultSqlDatabaseAndTable.Begin(), which causes things go badly.

**Comment by FriedrichWeinmann** [comment]
Alright, I was able to set up a repro and figure things out.
The debugging as done here turns out to not work as PowerShell expects it, so the ObjectNotFound part is caused by the debugging, not the original code.

I found the error with the original code, which happened in the compiled psm1 at line 15588:

```powershell
$queryParameters = $script:converter.Process($ObjectToProcess)
´´´

which should have been:

```powershell
$queryParameters = @($script:converter.Process($ObjectToProcess))[0]
```

I forgot that steppable pipelines return an array (of one item) in process.
There's a followup error because the default columns it tries to send have more columns than the default SQL table gets, which ... yeah, should not have happened.
I can only apologize for my sloppy work here - that should definitely not have made it through testing.

Will ship a fixed version tomorrow (including some extra tests to help prevent a regression).

**Comment by FriedrichWeinmann** [comment]
Alright, 1.7.245 should fix things

**Comment by ashdar** [comment]
Thanks for the quick turnaround.

I've tested 1.7.245 and it seems to connect appropriately and write a few log entries. After stumbling onto the fact the shape of the default sql table has changed, I can now easily log a hundred or so entries in less than a second.

I can now go back to my original problem, which is that I think that "($SqlInstance.ConnectionContext.IsOpen -ne 'True')" in the sql.provider.ps1 file is wrong because we are testing a boolean return value (IsOpen) against a string). If I need help with that, I will open a new item.

**Comment by ashdar** [comment]
You are right. It is 'wrong', but the consequence will be that it just tries to open the connection. I don't think that will cause the thing I am seeing. Basically, when I try to log 'lots' of things to a sql provider with something like:

        $LoopMax = 120
        foreach($i in @(1..$LoopMax)) {
            $Message = "Loop $i of $LoopMax"
            Write-PSFMessage -Message $Message -Level 'Verbose'
            Start-Sleep -Milliseconds 10
        }

I wind up with errors like:
      | Error connecting to [DEVSQLSERVER]: Timeout expired.  The timeout period elapsed prior to obtaining a connection from the pool.  This may have
      | occurred because all pooled connections were in use and max pool size was reached.

The number of loops I can get through seems to vary a bit depending on which VM/laptop/SQLServer I'm using, and I haven't had time to figure out any more details than this, which is why I haven't filed a bug on it yet. 

I had hoped that that simple IsOpen() thing that I pointed out was just causing psframework to keep asking dbatools for new connections until it ran the pool out, but maybe not. I hope that I can get back into the details on this in the next couple of days, **but this item (#546) does seem to be fixed.**

**Comment by FriedrichWeinmann** [comment]
Found the issue:
Basically, it will open a new connection each time if you specify the SQL Server by name, rather than offer an established connection, then never close them but open again and again and again.
Simple enough to fix, will try to make time for that tomorrow.

**Comment by FriedrichWeinmann** [comment]
Thank _you_ for poking at my logging system (which very much is the pride and flagship feature of my tool) and help me make it better :)

Implemented the promised fix and - for extra awesomeness (and some shame, admittedly) - it will now finally honor credentials.
Oh, and in case you never noticed or I forgot to document it: Did you know that the `-SqlServer` parameter of the logging provider also takes a pre-established connection?
So if you need some fancy custom settings, you can run Connect-DbaInstance in the main code and pass it to the logger (just make sure to not use the same connection in the main workflow - connections cannot be shared safely across runspaces).

Tests are running now and one this PR has been merged it (plus some minutes for release pipeline) the next version should be ready:
https://github.com/PowershellFrameworkCollective/psframework/pull/549

Please verify the new version on your end - I just have a demo local instance, so some field testing would be appreciated :)

**Comment by ashdar** [comment]
I updated a DEV machine and I ran what I consider to be a 'big test': Log one "started" entry, log 1200 "doing stuff now" entries and then log one final "completed" entry. It did take a few seconds beyond the end of my test script before all of the entries actually made it into the logging table (it's queuing, after all), but my test did create all 1202 log entries as I expected.  

I've got a couple of different projects (including things that will be in production real soon now) that I anticipate using PSFMessage for. I hope to have at least one of those deployed and running before the end of October. If any other quirk else pops up, I'll start another GitHub item. 

Over the past two years, I've written a couple of bad logging systems because I was too much in a hurry to research what was available. At this time, without PSFramework, I'd be looking at trying to get log4net working. A Powershell-based alternative seems easier to administrate and keep up to date than dealing with a nuget package and a DLL. 

As you know, PSFramework is a dependency to dbatools, which I've been using for many years. The main wins for me with PSFramework are being able to change logging targets by configuration, having the asynchronous queueing is sweet (even if debugging is not straightforward) and being able to log to a SQL table (because, in my environment, a SQL table is easier to search and maintain that than dealing with Windows Event Log on several different machines).

Thx!
